xain/grpc/participant.py

```diff
@@ -3,20 +3,35 @@
 from typing import Tuple
 
 import grpc
+from absl import app, flags
 from numproto import ndarray_to_proto, proto_to_ndarray
 
+from xain.benchmark.net import load_lr_fn_fn, load_model_fn
+from xain.datasets import load_splits
+from xain.fl.participant import ModelProvider, Participant
 from xain.grpc import coordinator_pb2, coordinator_pb2_grpc
 from xain.types import History, Metrics, Theta
 
+FLAGS = flags.FLAGS
+
 RETRY_TIMEOUT = 5
 HEARTBEAT_TIME = 10
 
 
-def heartbeat(channel, terminate_event):
+def heartbeat(channel, terminate_event, selected_event):
     stub = coordinator_pb2_grpc.CoordinatorStub(channel)
     while not terminate_event.is_set():
+        # exchange a heartbeat
         reply = stub.Heartbeat(coordinator_pb2.HeartbeatRequest())
         print(f"Participant received: {type(reply)}")
+        if reply.state == coordinator_pb2.State.FINISHED:
+            terminate_event.set()
+            return
+        if reply.state == coordinator_pb2.State.ROUND:
+            # signal "round open" to main thread
+            selected_event.set()
+        # not much to do for State.STANDBY
+        selected_event.clear()
```

Comment:
Here you are possibly setting and clearing the event very fast. I don't know if this is enough for the `train_when_selected` to pick it up. I think it would be safer for the event to be cleared inside `train_when_selected`.

Also this still needs logic to check the number of the round or else you may be doing multiple trainings per round.

Reply:
I thought about this too at the time - a quick search suggested it was fine, but I take your suggestion as an appropriate place to `clear()`.

Yes OK, because the Coordinator moves from ROUND i -> ROUND i+1. I had a slightly different mental model of the Coordinator so that the Participant could maybe get away with not tracking the round number. Maybe we can discuss later.